presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java

```diff
@@ -1033,6 +1034,15 @@ public void testNans()
         }
     }
 
+    @Test
+    public void testFilterFunctionsWithOptimization()
+    {
+        // the RowExpressionOptimizer optimizes the second part of the 2nd cardinality function to null hence not pushing down the linenumber as required predicate column
+        assertQuerySucceeds(getSession(), "SELECT partkey FROM lineitem WHERE " +
+                "cardinality(ARRAY_INTERSECT(ARRAY[orderkey], map_values(CAST(json_extract(json_parse('{\"a\": {\"a1\": {\"0\": 1}}}'),concat(concat('$.', 'a'), '[\"a1\"]')) AS MAP(VARCHAR, BIGINT))))) > 0 " +
```

Comment:
This query is a bit difficult to understand. Consider a simpler one:

```
SELECT partkey FROM lineitem WHERE orderkey > 10 OR if(json_extract(json_parse('{}'), '$.a') IS NOT NULL, quantity * discount) > 0
```

The results of that query can be verified and the verification could double-serve as documentation:

```
        assertQuery("SELECT partkey FROM lineitem WHERE orderkey > 10 OR if(json_extract(json_parse('{}'), '$.a') IS NOT NULL, quantity * discount) > 0",
                "SELECT partkey FROM lineitem WHERE orderkey > 10");
```

It would be nice to add this test to TestHiveLogicalPlanner where we can verify explicitly the list of columns in layout.predicateColumns.